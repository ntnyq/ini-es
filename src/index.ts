export * from './safe'
export * from './decode'
export * from './encode'
export * from './unsafe'
