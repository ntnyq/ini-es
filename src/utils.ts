import type { AnyObject } from './types'

export function hasOwn(obj: AnyObject, key: PropertyKey): boolean {
  return Object.prototype.hasOwnProperty.call(obj, key)
}

export function isQuoted(value: string): boolean {
  return (
    (value.startsWith(`"`) && value.endsWith(`"`))
    || (value.startsWith(`'`) && value.endsWith(`'`))
  )
}

export function splitSections(text: string, separator: string): string[] {
  let lastMatchIndex = 0
  let lastSeparatorIndex = 0
  let nextIndex = 0
  const sections: string[] = []

  do {
    nextIndex = text.indexOf(separator, lastMatchIndex)

    if (nextIndex !== -1) {
      lastMatchIndex = nextIndex + separator.length

      if (nextIndex > 0 && text[nextIndex - 1] === '\\') {
        continue
      }

      sections.push(text.slice(lastSeparatorIndex, nextIndex))
      lastSeparatorIndex = nextIndex + separator.length
    }
  } while (nextIndex !== -1)

  sections.push(text.slice(lastSeparatorIndex))

  return sections
}
