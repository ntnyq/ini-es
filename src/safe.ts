import { isQuoted } from './utils'

/**
 * Turn the given string into a safe to use key or value in your INI file
 *
 * @param text - given string
 * @returns init key or value
 */
export function safe(text: string): string {
  if (
    typeof text !== 'string'
    || /[=\r\n]/.test(text)
    || /^\[/.test(text)
    || (text.length > 1 && isQuoted(text))
    || text !== text.trim()
  ) {
    return JSON.stringify(text)
  }
  return text.split(';').join('\\;').split('#').join('\\#')
}
