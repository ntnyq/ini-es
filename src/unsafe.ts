import { isQuoted } from './utils'

/**
 * Revert `safe` processed result
 *
 * @param text - given ini key or value
 * @returns reversed safe process result
 */
export function unsafe(text: string): string {
  text = (text || '').trim()
  if (isQuoted(text)) {
    // remove the single quotes before calling JSON.parse
    if (text.charAt(0) === "'") {
      text = text.slice(1, -1)
    }
    try {
      text = JSON.parse(text)
    } catch {
      // ignore errors
    }
  } else {
    // walk the text to find the first not-escaped ; character
    let esc = false
    let unesc = ''
    for (let i = 0, l = text.length; i < l; i++) {
      const c = text.charAt(i)
      if (esc) {
        if ('\\;#'.includes(c)) {
          unesc += c
        } else {
          unesc += `\\${c}`
        }

        esc = false
      } else if (';#'.includes(c)) {
        break
      } else if (c === '\\') {
        esc = true
      } else {
        unesc += c
      }
    }
    if (esc) {
      unesc += '\\'
    }

    return unesc.trim()
  }
  return text
}
