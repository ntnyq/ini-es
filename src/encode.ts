import { safe } from './safe'
import { splitSections } from './utils'
import type { AnyObject } from './types'

export interface EncodeOptions {
  /**
   * Whether to align the `=` character for each section.
   *
   * -> Also enables the `whitespace` option
   *
   * @default false
   */
  align?: boolean

  /**
   * Whether to append `[]` to array keys.
   *
   * Some parsers treat duplicate names by themselves as arrays
   *
   * @default true
   */
  bracketedArray?: boolean

  /**
   * Whether to insert a newline after each section header.
   *
   * The TOSHIBA & FlashAir parser require this format.
   *
   * @default false
   */
  newline?: boolean

  /**
   * Which platforms line-endings should be used.
   *
   * win32 -> CR+LF
   * other -> LF
   *
   * Default is the current platform
   *
   * @default process.platform
   */
  platform?: Platform

  /**
   * Specify section to decode
   */
  section?: string

  /**
   * Whether to sort all sections & their keys alphabetically.
   *
   * @default false
   */
  sort?: boolean

  /**
   * Whether to insert spaces before & after `=`
   *
   * Disabled by default to have better
   * compatibility with old picky parsers.
   *
   * @default false
   */
  whitespace?: boolean
}

/**
 * Type of `process.platform`
 */
type Platform =
  | 'aix'
  | 'android'
  | 'cygwin'
  | 'darwin'
  | 'freebsd'
  | 'haiku'
  | 'linux'
  | 'netbsd'
  | 'openbsd'
  | 'sunos'
  | 'win32'

/**
 * Encodes the given data object as an INI formatted string
 *
 * @param obj - given data
 * @param options - encode options {@link EncodeOptions}
 * @returns formatted string
 */
export function encode(
  obj: AnyObject,
  options: string | EncodeOptions = {},
): string {
  if (typeof options === 'string') {
    options = {
      section: options,
    }
  }

  const {
    align = false,
    bracketedArray = true,
    newline = false,
    platform = typeof process == 'undefined' ? undefined : process.platform,
    section,
    sort = false,
    whitespace = options.align || false,
  } = options

  const eol = platform === 'win32' ? '\r\n' : '\n'
  const separator = whitespace ? ' = ' : '='
  const children: string[] = []

  const keys = sort ? Object.keys(obj).sort() : Object.keys(obj)

  let padToChars = 0
  // If aligning on the separator, then padToChars is determined as follows:
  // 1. Get the keys
  // 2. Exclude keys pointing to objects unless the value is null or an array
  // 3. Add `[]` to array keys
  // 4. Ensure non empty set of keys
  // 5. Reduce the set to the longest `safe` key
  // 6. Get the `safe` length
  if (align) {
    padToChars = safe(
      keys
        .filter(
          k =>
            obj[k] === null
            || Array.isArray(obj[k])
            || typeof obj[k] !== 'object',
        )
        .map(k => (Array.isArray(obj[k]) ? `${k}[]` : k))
        .concat([''])
        .reduce((a, b) => (safe(a).length >= safe(b).length ? a : b)),
    ).length
  }

  let out = ''
  const arraySuffix = bracketedArray ? '[]' : ''

  for (const k of keys) {
    const val = obj[k]
    if (val && Array.isArray(val)) {
      for (const item of val) {
        out +=
          safe(`${k}${arraySuffix}`).padEnd(padToChars, ' ')
          + separator
          + safe(item)
          + eol
      }
    } else if (val && typeof val === 'object') {
      children.push(k)
    } else {
      out += safe(k).padEnd(padToChars, ' ') + separator + safe(val) + eol
    }
  }

  if (section && out.length) {
    out = `[${safe(section)}]${newline ? eol + eol : eol}${out}`
  }

  for (const k of children) {
    const nk = splitSections(k, '.').join('\\.')
    const newSection = (section ? `${section}.` : '') + nk
    const child = encode(obj[k], {
      align,
      bracketedArray,
      newline,
      platform,
      section: newSection,
      sort,
      whitespace,
    })
    if (out.length && child.length) {
      out += eol
    }

    out += child
  }

  return out
}

/**
 * alias of `encode`
 *
 * @see {@link encode}
 */
export const stringify: typeof encode = encode
