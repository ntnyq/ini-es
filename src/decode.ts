import { unsafe } from './unsafe'
import { hasOwn, splitSections } from './utils'
import type { AnyObject } from './types'

export interface DecodeOptions {
  /**
   * Whether to append `[]` to array keys.
   * Some parsers treat duplicate names by themselves as arrays
   *
   * @default true
   */
  bracketedArray?: boolean
}

/**
 * Attempts to turn the given INI string into a nested data object.
 *
 * @param text - init string
 * @param options - decode options {@link DecodeOptions}
 * @returns nested data object
 */
export function decode<T extends AnyObject = AnyObject>(
  text: string,
  options: DecodeOptions = {},
): T {
  const { bracketedArray = true } = options
  const out: AnyObject = Object.create(null)
  let p: AnyObject = out
  let section = null

  //          section          |key      = value
  const re = /^\[([^\]]*)\]\s*$|^([^=]+)(=(.*))?$/
  const lines = text.split(/[\r\n]+/g)
  const duplicates: Record<string, any> = {}

  for (const line of lines) {
    if (!line || /^\s*[;#]/.test(line) || /^\s*$/.test(line)) {
      continue
    }
    const match = line.match(re)
    if (!match) {
      continue
    }
    if (match[1] !== undefined) {
      section = unsafe(match[1])
      if (section === '__proto__') {
        // not allowed
        // keep parsing the section, but don't attach it.
        p = Object.create(null)
        continue
      }
      p = out[section] = out[section] || Object.create(null)
      continue
    }
    const keyRaw = unsafe(match[2])
    let isArray: boolean
    if (bracketedArray) {
      isArray = keyRaw.length > 2 && keyRaw.slice(-2) === '[]'
    } else {
      duplicates[keyRaw] = (duplicates?.[keyRaw] || 0) + 1
      isArray = duplicates[keyRaw] > 1
    }
    const key = isArray && keyRaw.endsWith('[]') ? keyRaw.slice(0, -2) : keyRaw

    if (key === '__proto__') {
      continue
    }
    const valueRaw = match[3] ? unsafe(match[4]) : true
    const value =
      valueRaw === 'true' || valueRaw === 'false' || valueRaw === 'null'
        ? JSON.parse(valueRaw)
        : valueRaw

    // Convert keys with '[]' suffix to an array
    if (isArray) {
      if (!hasOwn(p, key)) {
        p[key] = []
      } else if (!Array.isArray(p[key])) {
        p[key] = [p[key]]
      }
    }

    // safeguard against resetting a previously defined
    // array by accidentally forgetting the brackets
    if (Array.isArray(p[key])) {
      p[key].push(value)
    } else {
      p[key] = value
    }
  }

  // {a:{y:1},"a.b":{x:2}} --> {a:{y:1,b:{x:2}}}
  // use a filter to return the keys that have to be deleted.
  const remove: string[] = []

  for (const k of Object.keys(out)) {
    if (
      !hasOwn(out, k)
      || typeof out[k] !== 'object'
      || Array.isArray(out[k])
    ) {
      continue
    }

    // see if the parent section is also an object.
    // if so, add it to that, and mark this one for deletion
    const parts = splitSections(k, '.')
    p = out
    const l = parts.pop() || ''
    const nl = l.replace(/\\\./g, '.')
    for (const part of parts) {
      if (part === '__proto__') {
        continue
      }
      if (!hasOwn(p, part) || typeof p[part] !== 'object') {
        p[part] = Object.create(null)
      }
      p = p[part]
    }
    if (p === out && nl === l) {
      continue
    }

    p[nl] = out[k]
    remove.push(k)
  }

  for (const del of remove) {
    delete out[del]
  }

  return out as T
}

/**
 * alias of `decode`
 *
 * @see {@link decode}
 */
export const parse: typeof decode = decode
