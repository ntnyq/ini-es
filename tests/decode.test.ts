import { expect, it } from 'vitest'
import { decode, parse } from '../src'
import { FIXTURE_DUPLICATED, FIXTURE_FOO } from './fixtures'

const parsedFoo = parse(FIXTURE_FOO)
const decodedFoo = decode(FIXTURE_FOO)

const parsedDuplicated = parse(FIXTURE_DUPLICATED)
const decodedDuplicated = parse(FIXTURE_DUPLICATED)

it('decode foo', () => {
  expect(decodedFoo).toMatchSnapshot()
})

it('decode duplicated', () => {
  expect(decodedDuplicated).toMatchSnapshot()
})

it('parse is alias of decode', () => {
  expect(parsedFoo).toStrictEqual(decodedFoo)
  expect(parsedDuplicated).toStrictEqual(decodedDuplicated)
})

it('decode foo - bracketedArray', () => {
  expect(
    decode(FIXTURE_FOO, {
      bracketedArray: false,
    }),
  ).toMatchSnapshot()
})
