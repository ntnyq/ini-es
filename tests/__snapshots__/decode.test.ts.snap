// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`decode duplicated 1`] = `
{
  "ar": [
    "one",
    "three",
  ],
  "brr": "3",
  "str": "3",
  "zr": [
    "deedee",
    "123",
  ],
}
`;

exports[`decode foo - bracketedArray 1`] = `
{
  " xa  n          p ": ""
yoyoyo

",
  "[disturbing]": "hey you never know",
  "a": {
    "[]": "a square?",
    "av": "a val",
    "b": {
      "c": {
        "e": [
          "1",
        ],
        "j": [
          "2",
        ],
      },
    },
    "cr": [
      "eight",
    ],
    "cr[]": "four",
    "e": "{ o: p, a: { av: a val, b: { c: { e: "this [value]" } } } }",
    "j": ""{ o: "p", a: { av: "a val", b: { c: { e: "this [value]" } } } }"",
  },
  "a with spaces": "b  c",
  "ar": [
    "three",
    "this is included",
  ],
  "ar[]": "one",
  "b": {},
  "br": [
    "cold",
    "warm",
  ],
  "eq": "eq=eq",
  "false": false,
  "null": null,
  "o": "p",
  "s": "something",
  "s1": ""something'",
  "s2": "something else",
  "s3": "",
  "s4": "",
  "s5": "   ",
  "s6": " a ",
  "s7": true,
  "true": true,
  "undefined": "undefined",
  "x.y.z": {
    "a.b.c": {
      "a.b.c": "abc",
      "noHashComment": "this# this is not a comment",
      "nocomment": "this; this is not a comment",
    },
    "x.y.z": "xyz",
  },
  "zr[]": "deedee",
}
`;

exports[`decode foo 1`] = `
{
  " xa  n          p ": ""
yoyoyo

",
  "[disturbing]": "hey you never know",
  "a": {
    "[]": "a square?",
    "av": "a val",
    "b": {
      "c": {
        "e": "1",
        "j": "2",
      },
    },
    "cr": [
      "four",
      "eight",
    ],
    "e": "{ o: p, a: { av: a val, b: { c: { e: "this [value]" } } } }",
    "j": ""{ o: "p", a: { av: "a val", b: { c: { e: "this [value]" } } } }"",
  },
  "a with spaces": "b  c",
  "ar": [
    "one",
    "three",
    "this is included",
  ],
  "b": {},
  "br": "warm",
  "eq": "eq=eq",
  "false": false,
  "null": null,
  "o": "p",
  "s": "something",
  "s1": ""something'",
  "s2": "something else",
  "s3": "",
  "s4": "",
  "s5": "   ",
  "s6": " a ",
  "s7": true,
  "true": true,
  "undefined": "undefined",
  "x.y.z": {
    "a.b.c": {
      "a.b.c": "abc",
      "noHashComment": "this# this is not a comment",
      "nocomment": "this; this is not a comment",
    },
    "x.y.z": "xyz",
  },
  "zr": [
    "deedee",
  ],
}
`;
